dynamic_rest/prefetch.py

```diff
@@ -0,0 +1,482 @@
+from collections import defaultdict
+import copy
+
+from django.db import models
+from django.db.models import Prefetch, QuerySet
+
+from dynamic_rest.meta import (
+    get_model_field_and_type,
+    get_remote_model,
+    reverse_m2m_field_name,
+    reverse_o2o_field_name
+)
+
+
+class FastObject(dict):
+
+    def __init__(self, *args, **kwargs):
+        self.pk_field = kwargs.pop('pk_field', 'id')
+        return super(FastObject, self).__init__(*args)
+
+    @property
+    def pk(self):
+        return self[self.pk_field]
+
+    def _slow_getattr(self, name):
+        if '.' in name:
+            parts = name.split('.')
+            obj = self
+            for part in parts:
+                obj = obj[part]
+            return obj
+        elif name == '*':
+            return self
+        else:
+            raise AttributeError(name)
+
+    def __getattr__(self, name):
+        try:
+            return self[name]
+        except KeyError:
+            # Fast approach failed, fall back on slower logic.
+            return self._slow_getattr(name)
+
+    def __setattr__(self, name, value):
+        if name != 'pk_field' and name != 'pk':
+            self[name] = value
+        else:
+            super(FastObject, self).__setattr__(name, value)
+
+
+class SlowObject(dict):
+
+    def __init__(self, slow_object=None, *args, **kwargs):
+        self.pk_field = kwargs.pop('pk_field', 'id')
+        self.data = slow_object
+        return super(SlowObject, self).__init__(slow_object.__dict__, *args)
+
+    @property
+    def pk(self):
+        return self[self.pk_field]
+
+    def __getitem__(self, value):
+
+        if hasattr(self.data, str(value)):
+            return getattr(self.data, str(value))
+
+        # for the purposse of mapping serialized model + '_id' fields back to
+        # internal models, we need to check if that pattern is present
+        is_nested_obj = value.split('_')
+        test_attr = '_'.join(is_nested_obj[:-1])
+        attr_exists = hasattr(self.data, test_attr)
+        if is_nested_obj[-1] == 'id' and attr_exists:
```

Comment:
Kind of curious about this... in what cases would `*_id` not be available? Doesn't Django automatically provide those for all single-relations?

Reply:
I'm guessing this would handle the case where the foreign key has a custom column name.